File: packages/nodx-js/src/package.mjs
import { parse } from "./blockParser.mjs";
import { sha256Base64Url } from "./bytes.mjs";
import { DEFAULT_LIMITS } from "./limits.mjs";
import { normalizePackagePath } from "./url.mjs";

const textDecoder = new TextDecoder("utf-8", { fatal: true });

export function isPackagedNodx(bytes) {
  return bytes.length >= 4 && bytes[0] === 0x50 && bytes[1] === 0x4b && bytes[2] === 0x03 && bytes[3] === 0x04;
}

export function openStoredPackage(bytes, limits = DEFAULT_LIMITS) {
  if (!(bytes instanceof Uint8Array)) bytes = new Uint8Array(bytes);
  if (!isPackagedNodx(bytes)) throw new Error("not a packaged NODX");
  const entries = zipEntries(bytes, limits);
  if (entries.length > limits.packageFileCount) throw new Error("package file count limit exceeded");
  const total = entries.reduce((sum, entry) => sum + entry.uncompressedSize, 0);
  if (total > limits.packageUncompressedBytes) throw new Error("package uncompressed size limit exceeded");
  if (entries[0]?.name !== "mimetype" || entries[0].compression !== 0) throw new Error("first ZIP entry must be mimetype");

  const files = new Map();
  for (const entry of entries) files.set(entry.name, readStoredEntry(bytes, entry, limits));

  const mimetype = files.get("mimetype");
  if (!mimetype || decode(mimetype) !== "application/nodx+zip") throw new Error("invalid NODX package mimetype");
  const manifestBytes = files.get("manifest.yaml");
  if (!manifestBytes) throw new Error("package is missing manifest.yaml");
  const manifest = parseManifest(decode(manifestBytes), limits);
  if (manifest.schema !== "nodx-package/1.0") throw new Error("package manifest schema must be nodx-package/1.0");
  if (manifest.entries.length > limits.manifestEntries) throw new Error("package manifest entry limit exceeded");
  for (const item of manifest.entries) {
    const data = files.get(item.path);
    if (!data) throw new Error("manifest lists a missing package entry");
    if (item.size !== null && item.size !== data.length) throw new Error("package manifest size does not match entry bytes");
    if (item.sha256 && item.sha256 !== sha256Base64Url(data)) throw new Error("package digest mismatch");
  }
  for (const path of [...manifestPaths(manifest.components), ...manifestPaths(manifest.themes)]) {
    if (!files.has(path)) throw new Error("manifest lists a missing package extension");
    if (!manifest.entries.some((entry) => entry.path === path)) throw new Error("package extension paths must also be listed in manifest entries");
  }
  if (!manifest.entry) throw new Error("package manifest is missing entry");
  manifest.entry = safePackagePath(manifest.entry, limits);
  if (!files.has(manifest.entry)) throw new Error("package entry document is missing");
  if (manifest.signature) {
    manifest.signature = safePackagePath(manifest.signature, limits);
    if (!files.has(manifest.signature)) throw new Error("package manifest references a missing signature");
  }
  return { entryPath: manifest.entry, entryBytes: files.get(manifest.entry), files, manifest };
}

export function packageEntryText(bytes) {
  return decode(openStoredPackage(bytes).entryBytes);
}

export function parsePackagedDocument(bytes, options = {}) {
  const pkg = openStoredPackage(bytes);
  const doc = parse(decode(pkg.entryBytes));
  return applyPackageExtensions(doc, pkg, options);
}

export function applyPackageExtensions(doc, pkg, options = {}) {
  const out = {
    ...doc,
    meta: {
      ...doc.meta,
      components: [...(Array.isArray(doc.meta.components) ? doc.meta.components : [])],
      stylesheets: [...(Array.isArray(doc.meta.stylesheets) ? doc.meta.stylesheets : [])],
    },
  };
  for (const component of packageComponentDefinitions(pkg, options)) out.meta.components.push(component);
  for (const css of packageThemeStylesheets(pkg, options)) out.meta.stylesheets.push(css);
  if (!out.meta.components.length) delete out.meta.components;
  if (!out.meta.stylesheets.length) delete out.meta.stylesheets;
  return out;
}

export function packageComponentDefinitions(pkg, options = {}) {
  const files = pkg.files;
  const paths = manifestPaths(pkg.manifest.components);
  if (options.autodiscoverComponents) {
    for (const path of files.keys()) {
      if (path.startsWith("components/") && path.endsWith(".nodx") && !paths.includes(path)) paths.push(path);
    }
  }
  return paths.map((path) => parseComponentFile(decodeRequired(files, path), path));
}

export function packageThemeStylesheets(pkg, options = {}) {
  const files = pkg.files;
  const paths = manifestPaths(pkg.manifest.themes);
  if (options.autodiscoverThemes) {
    for (const path of files.keys()) {
      if ((path.startsWith("themes/") || path.startsWith("styles/")) && path.endsWith(".nods") && !paths.includes(path)) paths.push(path);
    }
  }
  return paths.map((path) => decodeRequired(files, path));
}

function zipEntries(bytes, limits) {
  const eocd = findEocd(bytes);
  if (eocd < 0) throw new Error("ZIP end of central directory not found");
  if (u16(bytes, eocd + 4) !== 0 || u16(bytes, eocd + 6) !== 0) throw new Error("multi-disk ZIP is not supported");
  const count = u16(bytes, eocd + 10);
  const cdSize = u32(bytes, eocd + 12);
  const cdOffset = u32(bytes, eocd + 16);
  if (count === 0xffff || cdSize === 0xffffffff || cdOffset === 0xffffffff) throw new Error("ZIP64 packages are not supported");

  let pos = cdOffset;
  const entries = [];
  const seen = new Set();
  for (let i = 0; i < count; i += 1) {
    if (u32(bytes, pos) !== 0x02014b50) throw new Error("invalid ZIP central directory");
    const flags = u16(bytes, pos + 8);
    const compression = u16(bytes, pos + 10);
    const crc32 = u32(bytes, pos + 16);
    const compressedSize = u32(bytes, pos + 20);
    const uncompressedSize = u32(bytes, pos + 24);
    const nameLen = u16(bytes, pos + 28);
    const extraLen = u16(bytes, pos + 30);
    const commentLen = u16(bytes, pos + 32);
    const externalAttrs = u32(bytes, pos + 38);
    const localOffset = u32(bytes, pos + 42);
    if (compressedSize === 0xffffffff || uncompressedSize === 0xffffffff || localOffset === 0xffffffff) throw new Error("ZIP64 packages are not supported");
    if (uncompressedSize > limits.packageEntryBytes) throw new Error("package entry size limit exceeded");
    if (compressedSize === 0 && uncompressedSize > 0) throw new Error("package compression ratio limit exceeded");
    if (compressedSize > 0 && uncompressedSize > compressedSize * limits.packageCompressionRatio) throw new Error("package compression ratio limit exceeded");
    if (flags & 1) throw new Error("encrypted ZIP entries are not supported");
    if (flags & 8) throw new Error("ZIP data descriptors are not supported");
    if (compression !== 0) throw new Error("browser package reader supports stored ZIP entries only");
    rejectSpecialFile(externalAttrs);
    const nameStart = pos + 46;
    const nameEnd = nameStart + nameLen;
    const extraEnd = nameEnd + extraLen;
    if (extraEnd > bytes.length) throw new Error("ZIP entry name is out of bounds");
    rejectZip64Extra(bytes.subarray(nameEnd, extraEnd));
    const name = safePackagePath(decode(bytes.subarray(nameStart, nameEnd)), limits);
    if (seen.has(name)) throw new Error("duplicate package entry path");
    seen.add(name);
    entries.push({ name, compression, crc32, compressedSize, uncompressedSize, localOffset, externalAttrs });
    pos = extraEnd + commentLen;
  }
  return entries.sort((a, b) => a.localOffset - b.localOffset);
}

function readStoredEntry(bytes, entry, limits) {
  const offset = entry.localOffset;
  if (u32(bytes, offset) !== 0x04034b50) throw new Error("invalid ZIP local header");
  const flags = u16(bytes, offset + 6);
  const compression = u16(bytes, offset + 8);
  const crc = u32(bytes, offset + 14);
  const compressedSize = u32(bytes, offset + 18);
  const uncompressedSize = u32(bytes, offset + 22);
  if (flags & 1) throw new Error("encrypted ZIP entries are not supported");
  if (flags & 8) throw new Error("ZIP data descriptors are not supported");
  if (compression !== entry.compression || crc !== entry.crc32 || compressedSize !== entry.compressedSize || uncompressedSize !== entry.uncompressedSize) {
    throw new Error("ZIP local header does not match central directory");
  }
  const nameLen = u16(bytes, offset + 26);
  const extraLen = u16(bytes, offset + 28);
  const nameStart = offset + 30;
  const nameEnd = nameStart + nameLen;
  const extraEnd = nameEnd + extraLen;
  if (extraEnd > bytes.length) throw new Error("ZIP entry data is out of bounds");
  rejectZip64Extra(bytes.subarray(nameEnd, extraEnd));
  const name = safePackagePath(decode(bytes.subarray(nameStart, nameEnd)), limits);
  if (name !== entry.name) throw new Error("ZIP local header name mismatch");
  rejectSpecialFile(entry.externalAttrs);
  const end = extraEnd + entry.compressedSize;
  if (end > bytes.length) throw new Error("ZIP entry data is out of bounds");
  const data = bytes.slice(extraEnd, end);
  if (data.length !== entry.uncompressedSize) throw new Error("stored ZIP entry has inconsistent sizes");
  if (crc32(data) !== entry.crc32) throw new Error("ZIP CRC mismatch");
  if (looksLikeZip(data)) throw new Error("nested ZIP archives are not supported");
  return data;
}

function parseManifest(text, limits) {
  const manifest = { schema: "", entry: "", signature: "", entries: [], components: [], themes: [] };
  let section = "";
  let currentEntry = null;
  const seenTop = new Set();
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trimEnd();
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    if (!raw.startsWith(" ") && !raw.startsWith("-")) {
      section = "";
      currentEntry = null;
      const key = trimmed.endsWith(":") ? trimmed.slice(0, -1) : trimmed.split(":", 1)[0];
      if (seenTop.has(key)) throw new Error("duplicate manifest key");
      seenTop.add(key);
    }
    if (line.startsWith("schema:")) manifest.schema = unquote(line.slice(7).trim());
    else if (line.startsWith("entry:")) manifest.entry = unquote(line.slice(6).trim());
    else if (line.startsWith("signature:")) manifest.signature = unquote(line.slice(10).trim());
    else if (line === "entries:" || line === "components:" || line === "themes:") section = line.slice(0, -1);
    else if (section === "entries" && trimmed.startsWith("- path:")) {
      currentEntry = { path: safePackagePath(unquote(trimmed.slice(7).trim()), limits), size: null, sha256: "" };
      manifest.entries.push(currentEntry);
    } else if (section === "entries" && currentEntry && trimmed.startsWith("size:")) {
      currentEntry.size = Number(trimmed.slice(5).trim());
      if (!Number.isSafeInteger(currentEntry.size) || currentEntry.size < 0) throw new Error("invalid manifest entry size");
    } else if (section === "entries" && currentEntry && trimmed.startsWith("sha256:")) {
      currentEntry.sha256 = unquote(trimmed.slice(7).trim());
    } else if ((section === "components" || section === "themes") && trimmed.startsWith("- path:")) {
      manifest[section].push({ path: safePackagePath(unquote(trimmed.slice(7).trim()), limits) });
    }
  }
  return manifest;
}

function manifestPaths(items) {
  return Array.isArray(items) ? items.map((item) => typeof item === "string" ? item : item?.path).filter(Boolean) : [];
}

function parseComponentFile(text, path) {
  const body = stripFrontMatter(text);
  const parsed = parse(text);
  const name = typeof parsed.meta.name === "string" ? parsed.meta.name : basename(path).replace(/\.nodx$/, "");
  const component = { name, template: body };
  if (typeof parsed.meta.style === "string" && parsed.meta.style.trim()) component.style = parsed.meta.style;
  return component;
}

function stripFrontMatter(text) {
  if (!text.startsWith("---\n")) return text;
  const end = text.indexOf("\n---", 4);
  if (end < 0) return text;
  let bodyStart = end + 4;
  if (text[bodyStart] === "\r") bodyStart += 1;
  if (text[bodyStart] === "\n") bodyStart += 1;
  return text.slice(bodyStart);
}

function safePackagePath(path, limits) {
  const normalized = normalizePackagePath(path, limits);
  if (normalized === null) throw new Error("unsafe package path: " + path);
  return normalized;
}

function rejectZip64Extra(extra) {
  let pos = 0;
  while (pos + 4 <= extra.length) {
    const header = u16(extra, pos);
    const len = u16(extra, pos + 2);
    pos += 4;
    const end = pos + len;
    if (end > extra.length) throw new Error("invalid ZIP extra field");
    if (header === 0x0001) throw new Error("ZIP64 packages are not supported");
    pos = end;
  }
  if (pos !== extra.length) throw new Error("invalid ZIP extra field");
}

function rejectSpecialFile(externalAttrs) {
  const mode = externalAttrs >>> 16;
  if (mode === 0) return;
  if ((mode & 0o170000) !== 0o100000) throw new Error("ZIP special files are not supported");
}

function looksLikeZip(data) {
  return data.length >= 4 && data[0] === 0x50 && data[1] === 0x4b && (data[2] === 0x03 || data[2] === 0x05 || data[2] === 0x07);
}

function decodeRequired(files, path) {
  const bytes = files.get(path);
  if (!bytes) throw new Error("package manifest references missing path: " + path);
  return decode(bytes);
}

function basename(path) {
  return path.split("/").pop() ?? path;
}

function unquote(value) {
  if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) return value.slice(1, -1);
  return value;
}

function findEocd(bytes) {
  for (let i = bytes.length - 22; i >= Math.max(0, bytes.length - 65557); i -= 1) {
    if (u32(bytes, i) === 0x06054b50) return i;
  }
  return -1;
}

function decode(bytes) {
  return textDecoder.decode(bytes);
}

function u16(bytes, pos) {
  if (pos + 2 > bytes.length) throw new Error("unexpected end of ZIP data");
  return bytes[pos] | (bytes[pos + 1] << 8);
}

function u32(bytes, pos) {
  if (pos + 4 > bytes.length) throw new Error("unexpected end of ZIP data");
  return (bytes[pos] | (bytes[pos + 1] << 8) | (bytes[pos + 2] << 16) | (bytes[pos + 3] << 24)) >>> 0;
}

function crc32(bytes) {
  let crc = 0xffffffff;
  for (const byte of bytes) crc = (crc >>> 8) ^ CRC_TABLE[(crc ^ byte) & 0xff];
  return (crc ^ 0xffffffff) >>> 0;
}

const CRC_TABLE = new Uint32Array(256);
for (let i = 0; i < 256; i += 1) {
  let c = i;
  for (let k = 0; k < 8; k += 1) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
  CRC_TABLE[i] = c >>> 0;
}

File: packages/nodx-py/src/nodx/package.py
from dataclasses import dataclass
from zlib import crc32

from .block_parser import parse
from .bytes import sha256_base64_url
from .limits import DEFAULT_LIMITS
from .package_diagnostics import raise_diag
from .url import normalize_package_path


@dataclass(frozen=True)
class StoredPackage:
    entry_path: str
    entry_bytes: bytes
    files: dict
    manifest: dict


@dataclass(frozen=True)
class ZipEntry:
    name: str
    compression: int
    crc32: int
    compressed_size: int
    uncompressed_size: int
    local_offset: int
    external_attrs: int


def is_packaged_nodx(bytes_):
    return len(bytes_) >= 4 and bytes_[0:4] == b"PK\x03\x04"


def open_stored_package(bytes_, limits=None):
    limits = limits or DEFAULT_LIMITS
    bytes_ = bytes(bytes_)
    if not is_packaged_nodx(bytes_):
        raise_diag("NODX-E012", "Not a packaged NODX.")
    entries = zip_entries(bytes_, limits)
    if len(entries) > limits["packageFileCount"]:
        raise_diag("NODX-E012", "Package file count limit exceeded.")
    if sum(entry.uncompressed_size for entry in entries) > limits["packageUncompressedBytes"]:
        raise_diag("NODX-E012", "Package uncompressed size limit exceeded.")
    if not entries or entries[0].name != "mimetype" or entries[0].compression != 0:
        raise_diag("NODX-E012", "First ZIP entry must be mimetype.")

    files = {entry.name: read_stored_entry(bytes_, entry, limits) for entry in entries}
    mimetype = files.get("mimetype")
    if not mimetype or decode(mimetype) != "application/nodx+zip":
        raise_diag("NODX-E012", "Invalid NODX package mimetype.")
    manifest_bytes = files.get("manifest.yaml")
    if not manifest_bytes:
        raise_diag("NODX-E012", "Package is missing manifest.yaml.")
    manifest = parse_manifest(decode(manifest_bytes), limits)
    if manifest.get("schema") != "nodx-package/1.0":
        raise_diag("NODX-E012", "Package manifest schema must be `nodx-package/1.0`.")
    if len(manifest["entries"]) > limits["manifestEntries"]:
        raise_diag("NODX-E012", "Package manifest entry limit exceeded.")
    for item in manifest["entries"]:
        data = files.get(item["path"])
        if data is None:
            raise_diag("NODX-E021", "Manifest lists a missing package entry.")
        if item.get("size") is not None and item["size"] != len(data):
            raise_diag("NODX-E021", "Package manifest size does not match entry bytes.")
        if item.get("sha256") and item["sha256"] != sha256_base64_url(data):
            raise_diag("NODX-E021", "Package digest mismatch.")
    for path in manifest_paths(manifest.get("components")) + manifest_paths(manifest.get("themes")):
        if path not in files:
            raise_diag("NODX-E021", "Manifest lists a missing package extension.")
        if not any(entry["path"] == path for entry in manifest["entries"]):
            raise_diag("NODX-E021", "Package extension paths must also be listed in manifest entries.")
    if not manifest.get("entry"):
        raise_diag("NODX-E012", "Package manifest is missing entry.")
    manifest["entry"] = safe_package_path(manifest["entry"], limits)
    if manifest["entry"] not in files:
        raise_diag("NODX-E012", "Package entry document is missing.")
    if manifest.get("signature"):
        manifest["signature"] = safe_package_path(manifest["signature"], limits)
        if manifest["signature"] not in files:
            raise_diag("NODX-E012", "Package manifest references a missing signature.")
    return StoredPackage(manifest["entry"], files[manifest["entry"]], files, manifest)


def package_entry_text(bytes_):
    return decode(open_stored_package(bytes_).entry_bytes)


def parse_packaged_document(bytes_, options=None):
    pkg = open_stored_package(bytes_)
    doc = parse(decode(pkg.entry_bytes))
    return apply_package_extensions(doc, pkg, options)


def apply_package_extensions(doc, pkg, options=None):
    options = options or {}
    out = {**doc, "meta": {**doc["meta"]}}
    out["meta"]["components"] = list(out["meta"].get("components") or [])
    out["meta"]["stylesheets"] = list(out["meta"].get("stylesheets") or [])
    out["meta"]["components"].extend(package_component_definitions(pkg, options))
    out["meta"]["stylesheets"].extend(package_theme_stylesheets(pkg, options))
    if not out["meta"]["components"]:
        del out["meta"]["components"]
    if not out["meta"]["stylesheets"]:
        del out["meta"]["stylesheets"]
    return out


def package_component_definitions(pkg, options=None):
    options = options or {}
    paths = manifest_paths(pkg.manifest.get("components"))
    if options.get("autodiscover_components"):
        for path in pkg.files:
            if path.startswith("components/") and path.endswith(".nodx") and path not in paths:
                paths.append(path)
    return [parse_component_file(decode_required(pkg.files, path), path) for path in paths]


def package_theme_stylesheets(pkg, options=None):
    options = options or {}
    paths = manifest_paths(pkg.manifest.get("themes"))
    if options.get("autodiscover_themes"):
        for path in pkg.files:
            if (path.startswith("themes/") or path.startswith("styles/")) and path.endswith(".nods") and path not in paths:
                paths.append(path)
    return [decode_required(pkg.files, path) for path in paths]


def zip_entries(bytes_, limits):
    eocd = find_eocd(bytes_)
    if eocd < 0:
        raise_diag("NODX-E012", "ZIP end of central directory not found.")
    if u16(bytes_, eocd + 4) != 0 or u16(bytes_, eocd + 6) != 0:
        raise_diag("NODX-E012", "Multi-disk ZIP is not supported.")
    count = u16(bytes_, eocd + 10)
    cd_size = u32(bytes_, eocd + 12)
    cd_offset = u32(bytes_, eocd + 16)
    if count == 0xFFFF or cd_size == 0xFFFFFFFF or cd_offset == 0xFFFFFFFF:
        raise_diag("NODX-E012", "ZIP64 packages are not supported.")

    pos = cd_offset
    entries = []
    seen = set()
    for _ in range(count):
        if u32(bytes_, pos) != 0x02014B50:
            raise_diag("NODX-E012", "Invalid ZIP central directory.")
        flags = u16(bytes_, pos + 8)
        compression = u16(bytes_, pos + 10)
        entry_crc = u32(bytes_, pos + 16)
        compressed_size = u32(bytes_, pos + 20)
        uncompressed_size = u32(bytes_, pos + 24)
        name_len = u16(bytes_, pos + 28)
        extra_len = u16(bytes_, pos + 30)
        comment_len = u16(bytes_, pos + 32)
        external_attrs = u32(bytes_, pos + 38)
        local_offset = u32(bytes_, pos + 42)
        if compressed_size == 0xFFFFFFFF or uncompressed_size == 0xFFFFFFFF or local_offset == 0xFFFFFFFF:
            raise_diag("NODX-E012", "ZIP64 packages are not supported.")
        if uncompressed_size > limits["packageEntryBytes"]:
            raise_diag("NODX-E012", "Package entry size limit exceeded.")
        if compressed_size == 0 and uncompressed_size > 0:
            raise_diag("NODX-E012", "Package compression ratio limit exceeded.")
        if compressed_size > 0 and uncompressed_size > compressed_size * limits["packageCompressionRatio"]:
            raise_diag("NODX-E012", "Package compression ratio limit exceeded.")
        if flags & 1:
            raise_diag("NODX-E012", "Encrypted ZIP entries are not supported.")
        if flags & 8:
            raise_diag("NODX-E012", "ZIP data descriptors are not supported.")
        if compression != 0:
            raise_diag("NODX-E012", "Browser package reader supports stored ZIP entries only.")
        reject_special_file(external_attrs)
        name_start = pos + 46
        name_end = name_start + name_len
        extra_end = name_end + extra_len
        if extra_end > len(bytes_):
            raise_diag("NODX-E012", "ZIP entry name is out of bounds.")
        reject_zip64_extra(bytes_[name_end:extra_end])
        name = safe_package_path(decode(bytes_[name_start:name_end]), limits)
        if name in seen:
            raise_diag("NODX-E012", "Duplicate package entry path.")
        seen.add(name)
        entries.append(ZipEntry(name, compression, entry_crc, compressed_size, uncompressed_size, local_offset, external_attrs))
        pos = extra_end + comment_len
    return sorted(entries, key=lambda entry: entry.local_offset)


def read_stored_entry(bytes_, entry, limits):
    pos = entry.local_offset
    if u32(bytes_, pos) != 0x04034B50:
        raise_diag("NODX-E012", "Invalid ZIP local header.")
    flags = u16(bytes_, pos + 6)
    compression = u16(bytes_, pos + 8)
    entry_crc = u32(bytes_, pos + 14)
    compressed_size = u32(bytes_, pos + 18)
    uncompressed_size = u32(bytes_, pos + 22)
    if flags & 1:
        raise_diag("NODX-E012", "Encrypted ZIP entries are not supported.")
    if flags & 8:
        raise_diag("NODX-E012", "ZIP data descriptors are not supported.")
    if (
        compression != entry.compression
        or entry_crc != entry.crc32
        or compressed_size != entry.compressed_size
        or uncompressed_size != entry.uncompressed_size
    ):
        raise_diag("NODX-E012", "ZIP local header does not match central directory.")
    name_len = u16(bytes_, pos + 26)
    extra_len = u16(bytes_, pos + 28)
    name_start = pos + 30
    name_end = name_start + name_len
    extra_end = name_end + extra_len
    if extra_end > len(bytes_):
        raise_diag("NODX-E012", "ZIP entry data is out of bounds.")
    reject_zip64_extra(bytes_[name_end:extra_end])
    name = safe_package_path(decode(bytes_[name_start:name_end]), limits)
    if name != entry.name:
        raise_diag("NODX-E012", "ZIP local header name mismatch.")
    reject_special_file(entry.external_attrs)
    end = extra_end + entry.compressed_size
    if end > len(bytes_):
        raise_diag("NODX-E012", "ZIP entry data is out of bounds.")
    data = bytes_[extra_end:end]
    if len(data) != entry.uncompressed_size:
        raise_diag("NODX-E012", "Stored ZIP entry has inconsistent sizes.")
    if crc32(data) & 0xFFFFFFFF != entry.crc32:
        raise_diag("NODX-E012", "ZIP CRC mismatch.")
    if looks_like_zip(data):
        raise_diag("NODX-E010", "Nested ZIP archives are not supported.")
    return data


def parse_manifest(text, limits):
    manifest = {"schema": "", "entry": "", "signature": "", "entries": [], "components": [], "themes": []}
    section = ""
    current_entry = None
    seen_top = set()
    for raw in text.replace("\r\n", "\n").split("\n"):
        line = raw.rstrip()
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#"):
            continue
        if not raw.startswith(" ") and not raw.startswith("-"):
            section = ""
            current_entry = None
            key = trimmed[:-1] if trimmed.endswith(":") else trimmed.split(":", 1)[0]
            if key in seen_top:
                raise_diag("NODX-E012", "Duplicate manifest key.")
            seen_top.add(key)
        if line.startswith("schema:"):
            manifest["schema"] = unquote(line[7:].strip())
        elif line.startswith("entry:"):
            manifest["entry"] = unquote(line[6:].strip())
        elif line.startswith("signature:"):
            manifest["signature"] = unquote(line[10:].strip())
        elif line in ("entries:", "components:", "themes:"):
            section = line[:-1]
        elif section == "entries" and trimmed.startswith("- path:"):
            current_entry = {"path": safe_package_path(unquote(trimmed[7:].strip()), limits), "size": None, "sha256": ""}
            manifest["entries"].append(current_entry)
        elif section == "entries" and current_entry is not None and trimmed.startswith("size:"):
            current_entry["size"] = int(trimmed[5:].strip())
            if current_entry["size"] < 0:
                raise_diag("NODX-E012", "Invalid manifest entry size.")
        elif section == "entries" and current_entry is not None and trimmed.startswith("sha256:"):
            current_entry["sha256"] = unquote(trimmed[7:].strip())
        elif section in ("components", "themes") and trimmed.startswith("- path:"):
            manifest[section].append({"path": safe_package_path(unquote(trimmed[7:].strip()), limits)})
    return manifest


def manifest_paths(items):
    return [item if isinstance(item, str) else item.get("path") for item in items or [] if (item if isinstance(item, str) else item.get("path"))]


def parse_component_file(text, path):
    body = strip_front_matter(text)
    parsed = parse(text)
    name = parsed["meta"].get("name") if isinstance(parsed["meta"].get("name"), str) else path.rsplit("/", 1)[-1].removesuffix(".nodx")
    component = {"name": name, "template": body}
    if isinstance(parsed["meta"].get("style"), str) and parsed["meta"]["style"].strip():
        component["style"] = parsed["meta"]["style"]
    return component


def strip_front_matter(text):
    if not text.startswith("---\n"):
        return text
    end = text.find("\n---", 4)
    if end < 0:
        return text
    body_start = end + 4
    if body_start < len(text) and text[body_start] == "\r":
        body_start += 1
    if body_start < len(text) and text[body_start] == "\n":
        body_start += 1
    return text[body_start:]


def safe_package_path(path, limits):
    normalized = normalize_package_path(path, limits)
    if normalized is None:
        raise_diag("NODX-E010", f"Unsafe package path: {path}")
    return normalized


def reject_zip64_extra(extra):
    pos = 0
    while pos + 4 <= len(extra):
        header = int.from_bytes(extra[pos:pos + 2], "little")
        size = int.from_bytes(extra[pos + 2:pos + 4], "little")
        pos += 4
        end = pos + size
        if end > len(extra):
            raise_diag("NODX-E012", "Invalid ZIP extra field.")
        if header == 0x0001:
            raise_diag("NODX-E012", "ZIP64 packages are not supported.")
        pos = end
    if pos != len(extra):
        raise_diag("NODX-E012", "Invalid ZIP extra field.")


def reject_special_file(external_attrs):
    mode = external_attrs >> 16
    if mode == 0:
        return
    if mode & 0o170000 != 0o100000:
        raise_diag("NODX-E012", "ZIP special files are not supported.")


def looks_like_zip(data):
    return len(data) >= 4 and data[0:2] == b"PK" and data[2] in (0x03, 0x05, 0x07)


def find_eocd(bytes_):
    start = max(0, len(bytes_) - 65557)
    for pos in range(len(bytes_) - 22, start - 1, -1):
        if bytes_[pos:pos + 4] == b"PK\x05\x06":
            return pos
    return -1


def decode_required(files, path):
    if path not in files:
        raise_diag("NODX-E021", f"Package manifest references missing path: {path}")
    return decode(files[path])


def unquote(value):
    if (value.startswith('"') and value.endswith('"')) or (value.startswith("'") and value.endswith("'")):
        return value[1:-1]
    return value


def u16(bytes_, pos):
    if pos + 2 > len(bytes_):
        raise_diag("NODX-E012", "Unexpected end of ZIP data.")
    return int.from_bytes(bytes_[pos:pos + 2], "little")


def u32(bytes_, pos):
    if pos + 4 > len(bytes_):
        raise_diag("NODX-E012", "Unexpected end of ZIP data.")
    return int.from_bytes(bytes_[pos:pos + 4], "little")


def decode(bytes_):
    return bytes_.decode("utf-8", errors="strict")


isPackagedNodx = is_packaged_nodx
openStoredPackage = open_stored_package
packageEntryText = package_entry_text
parsePackagedDocument = parse_packaged_document
applyPackageExtensions = apply_package_extensions
packageComponentDefinitions = package_component_definitions
packageThemeStylesheets = package_theme_stylesheets
